File: Domains/InfiniteMario/15-rl-competition-2009/system/codecs/Java/src/org/rlcommunity/rlglue/codec/network/ClientAgent.java
/* 
 * Copyright (C) 2007, Brian Tanner
 *
http://rl-glue-ext.googlecode.com/

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 *
 *  $Revision: 638 $
 *  $Date: 2009-02-07 16:17:29 -0500 (Sat, 07 Feb 2009) $
 *  $Author: brian@tannerpages.com $
 *  $HeadURL: http://rl-glue-ext.googlecode.com/svn/trunk/projects/codecs/Java/src/org/rlcommunity/rlglue/codec/network/ClientAgent.java $
 *
 */
package org.rlcommunity.rlglue.codec.network;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import org.rlcommunity.rlglue.codec.AgentInterface;

import org.rlcommunity.rlglue.codec.types.Action;
import org.rlcommunity.rlglue.codec.types.Observation;

public class ClientAgent {

    protected static final String kUnknownMessage = "Unknown Message: ";
    protected Network network;
    protected AgentInterface agent;
    protected boolean killedFromLocalProcess = false;
    private boolean debug = false;

    /**
     *If you are using ClientAgent in a local context (like from Matlab)
     * this allows us to kill the agent without quitting matlab
     **/
    public void killProcess() {
        killedFromLocalProcess = true;
    }

    public ClientAgent(AgentInterface agent) {
        this.agent = agent;
        assert agent != null : "agent null in ClientAgent constructor";
        this.network = new Network();
    }

    protected void onAgentInit() throws UnsupportedEncodingException {
        String taskSpec = network.getString();

        agent.agent_init(taskSpec);

        network.clearSendBuffer();
        network.putInt(Network.kAgentInit);
        //BTANNER: Sept 16 2008.. why are we putting this here?
        //OH, it's the size!  Ok.  I think.
        network.putInt(0); // No data following this header
    }

    protected void onAgentStart() {
        if (debug) {
            System.out.println("\tonAgentStart()");
        }
        Observation observation = network.getObservation();
        if (debug) {
            System.out.println("\t\tgot observation");
        }
        Action action = agent.agent_start(observation);
        if (debug) {
            System.out.println("\t\tgot action");
        }

        int size = Network.sizeOf(action);

        network.clearSendBuffer();
        network.putInt(Network.kAgentStart);
        network.putInt(size);
        network.putAction(action);
    }

    protected void onAgentStep() {
        double reward = network.getDouble();
        Observation observation = network.getObservation();
        Action action = agent.agent_step(reward, observation);

        int size = Network.sizeOf(action);
        network.clearSendBuffer();
        network.putInt(Network.kAgentStep);
        network.putInt(size);
        network.putAction(action);
    }

    protected void onAgentEnd() {
        //int size = network.getInt();
        double reward = network.getDouble();

        agent.agent_end(reward);

        network.clearSendBuffer();
        network.putInt(Network.kAgentEnd);
        network.putInt(0); // No data in this packet
    }

    protected void onAgentCleanup() {
        agent.agent_cleanup();

        network.clearSendBuffer();
        network.putInt(Network.kAgentCleanup);
        network.putInt(0); // No data in this packet
    }

    protected void onAgentMessage() throws UnsupportedEncodingException {
        String message = network.getString();
        String reply = agent.agent_message(message);

        network.clearSendBuffer();
        network.putInt(Network.kAgentMessage);
        network.putInt(Network.sizeOf(reply));
        network.putString(reply);
    }

    /**
     * Called by agentloader.
     * @param host
     * @param port
     * @param timeout
     * @throws java.lang.Exception
     */
    public void connect(String host, int port, int timeout) throws Exception {
        network.connect(host, port, timeout);
        network.clearSendBuffer();
        network.putInt(Network.kAgentConnection);
        network.putInt(0); // No body to this packet
        network.flipSendBuffer();
        network.send();
    }

    public void close() throws IOException {
        network.close();
    }

    public void runAgentEventLoop() throws Exception {
        int agentState = 0;
        int dataSize = 0;
        int recvSize = 0;
        int remaining = 0;

        do {

            network.clearRecvBuffer();


            try {
                recvSize = network.recv(8) - 8; // We may have received the header and part of the payload
                // We need to keep track of how much of the payload was recv'd


                agentState = network.getInt(0);
                dataSize = network.getInt(Network.kIntSize);

                remaining = dataSize - recvSize;
                if (remaining < 0) {
                    System.out.println("Remaining was less than 0!");
                    remaining = 0;
                }
            } catch (RLGlueDisconnectException e) {
                System.err.println(e.getMessage());
               System.exit(1);
            }


            int amountReceived = network.recv(remaining);

            network.flipRecvBuffer();

            // We have already received the header, now we need to discard it.
            network.getInt();
            network.getInt();

            switch (agentState) {
                case Network.kAgentInit:
                    onAgentInit();
                    break;

                case Network.kAgentStart:
                    onAgentStart();
                    break;

                case Network.kAgentStep:
                    onAgentStep();
                    break;

                case Network.kAgentEnd:
                    onAgentEnd();
                    break;

                case Network.kAgentCleanup:
                    onAgentCleanup();
                    break;


                case Network.kAgentMessage:
                    onAgentMessage();
                    break;

                case Network.kRLTerm:
                    break;

                default:
                    System.err.println(kUnknownMessage + agentState);
                   System.exit(1);
                    break;
            }
            ;

            network.flipSendBuffer();
            network.send();
        } while (agentState != Network.kRLTerm && !killedFromLocalProcess);
    }
}

File: Domains/InfiniteMario/15-rl-competition-2009/system/codecs/Python/src/rlglue/agent/ClientAgent.py
# 
# Copyright (C) 2007, Mark Lee
# 
#http://rl-glue-ext.googlecode.com/
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
#
#  $Revision: 446 $
#  $Date: 2009-01-22 22:20:21 -0500 (Thu, 22 Jan 2009) $
#  $Author: brian@tannerpages.com $
#  $HeadURL: http://rl-glue-ext.googlecode.com/svn/trunk/projects/codecs/Python/src/rlglue/agent/ClientAgent.py $

import sys

import rlglue.network.Network as Network
from rlglue.types import Action
from rlglue.types import Observation

class ClientAgent:
	kUnknownMessage = "Unknown Message: "
	network = None
	agent = None

	# (agent) -> void
	def __init__(self, agent):
		self.agent = agent
		self.network = Network.Network()

	# () -> void
	def onAgentInit(self):
		taskSpec = self.network.getString()
		self.agent.agent_init(taskSpec)
		self.network.clearSendBuffer()
		self.network.putInt(Network.kAgentInit)
		self.network.putInt(0) # No data following this header

	# () -> void
	def onAgentStart(self):
		observation = self.network.getObservation()
		action = self.agent.agent_start(observation)
		size = self.network.sizeOfAction(action)
		self.network.clearSendBuffer()
		self.network.putInt(Network.kAgentStart)
		self.network.putInt(size)
		self.network.putAction(action)

	# () -> void
	def onAgentStep(self):
		reward = self.network.getDouble()
		observation = self.network.getObservation()
		action = self.agent.agent_step(reward, observation)
		size = self.network.sizeOfAction(action)
		self.network.clearSendBuffer()
		self.network.putInt(Network.kAgentStep)
		self.network.putInt(size)
		self.network.putAction(action)

	# () -> void
	def onAgentEnd(self):
		reward = self.network.getDouble()
		self.agent.agent_end(reward)
		self.network.clearSendBuffer()
		self.network.putInt(Network.kAgentEnd)
		self.network.putInt(0) # No data in this packet

	# () -> void
	def onAgentCleanup(self):
		self.agent.agent_cleanup()
		self.network.clearSendBuffer()
		self.network.putInt(Network.kAgentCleanup)
		self.network.putInt(0) # No data in this packet

	# () -> void
	def onAgentMessage(self):
		message = self.network.getString()
		reply = self.agent.agent_message(message)
		self.network.clearSendBuffer()
		self.network.putInt(Network.kAgentMessage)
		if reply == None:
			#Brian Tanner added payload even for empty message (IE: send that the size is 0)
			self.network.putInt(4)
			self.network.putInt(0)
		else:
			#Brian Tanner, added 4 to the payload size because we putString sends the string AND its size
			self.network.putInt(len(reply) + 4)
			self.network.putString(reply)

	# (string, int, int) -> void
	def connect(self, host, port, timeout):
		self.network.connect(host, port, timeout);
		self.network.clearSendBuffer()
		self.network.putInt(Network.kAgentConnection)
		self.network.putInt(0) # No body to this packet
		self.network.send()

	# () -> void
	def close(self):
		self.network.close()

	# () -> void
	def runAgentEventLoop(self):
		agentState = 0
		dataSize = 0
		recvSize = 0
		remaining = 0

		while agentState != Network.kRLTerm:
			self.network.clearRecvBuffer();
			recvSize = self.network.recv(8) - 8; # We may have received the header and part of the payload
											# We need to keep track of how much of the payload was recv'd
			agentState = self.network.getInt()
			dataSize = self.network.getInt()
			
			remaining = dataSize - recvSize;
			if (remaining < 0):
				print("Remaining was less than 0!")
				remaining = 0

			amountReceived = self.network.recv(remaining)
			
			# Already read the header, discard it
			self.network.getInt()
			self.network.getInt()

			switch = {
				Network.kAgentInit: lambda self: self.onAgentInit(),
				Network.kAgentStart: lambda self: self.onAgentStart(),
				Network.kAgentStep: lambda self: self.onAgentStep(),
				Network.kAgentEnd: lambda self: self.onAgentEnd(),
				Network.kAgentCleanup: lambda self: self.onAgentCleanup(),
				Network.kAgentMessage: lambda self: self.onAgentMessage() }
			if agentState in switch:
				switch[agentState](self)
			elif agentState == Network.kRLTerm:
				pass
			else:
				sys.stderr.write(Network.kUnknownMessage % (str(agentState)))
				sys.exit(1)

			self.network.send()
